File: mining-go/mining.go
package main

import (
	"fmt"
	"time"
	"crypto/sha256"
	"sync"
)

//  Difficulty of mining
var difficulty int = 5
var prefix string

//  Number of trials
var ntrial int = 5

var data string = "parallel-mining-benchmark"
var global_result string

func sequential() {
	for i := 0; i < ntrial; i++{
		ldata := data + string(i)
		nonce := 0
		for {
			sha256 := sha256.New()
			sha256.Write([]byte(ldata + string(nonce)))
			hash := fmt.Sprintf("%x", sha256.Sum(nil))
			if hash[0:difficulty] == prefix {
				global_result = hash
				break
			}
			nonce++
		}
	}
}

var found bool = false
func parallelGoroutine(ngoroutine int) {
	for i := 0; i < ntrial; i++ {
		var wg sync.WaitGroup
		wg.Add(ngoroutine)
		ldata := data + string(i)
		found = false
		for j := 0; j < ngoroutine; j++ {
			// go mining(j, ngoroutine, ldata, wg)
			go func(id int) {
				defer wg.Done()
				nonce := id
				for !found {
					sha256 := sha256.New()
					sha256.Write([]byte(ldata + string(nonce)))
					hash := fmt.Sprintf("%x", sha256.Sum(nil))
					if hash[0:difficulty] == prefix {
						global_result = hash
						found = true
					}
					nonce += ngoroutine
				}
			} (j)
		}

		wg.Wait()
	}
}

func main() {
	for i := 0; i < difficulty; i++ {
		prefix += "0"
	}
	line := ""
	for i := 0; i < 40; i++ {
		line += "-"
	}
	fmt.Println(line)
	fmt.Println("Golang Version:")
    fmt.Printf("difficulty: %v, number of trials: %v\n", difficulty, ntrial)

	// sequential
	fmt.Println(line)
	start := time.Now()
	sequential()
	elapsed := time.Since(start)
	fmt.Println("sequential version:")
	fmt.Printf("total time consumed: %.4f s\n", elapsed.Seconds())
	fmt.Printf("average time consumed: %.4f s\n", elapsed.Seconds() / float64(ntrial))
	fmt.Printf("last hash computed: %v\n", global_result)

	// goroutine parallel
	ngoroutines := []int{2, 4, 8, 16}
	for _, ngoroutine := range ngoroutines {
		fmt.Println(line)
		start := time.Now()
		parallelGoroutine(ngoroutine)
		elapsed := time.Since(start)
		fmt.Printf("%v goroutine version:\n", ngoroutine)
		fmt.Printf("total time consumed: %.4f s\n", elapsed.Seconds())
		fmt.Printf("average time consumed: %.4f s\n", elapsed.Seconds() / float64(ntrial))
		fmt.Printf("last hash computed: %v\n", global_result)
	}
}

File: mining-python/mining.py
import hashlib
import time
import threading
import multiprocessing

# Difficulty of mining
difficulty = 5
prefix = "0" * difficulty

# Number of trials
ntrial = 5

data = "parallel-mining-benchmark"
global_result = None

def sequential():
    nonce = 0
    global global_result
    for i in range(ntrial):
        ldata = data + str(i)
        while True:
            s = ldata + str(nonce)
            result = hashlib.sha256(s.encode())
            if result.hexdigest()[:5] == prefix:
                # print(result.hexdigest())
                global_result = result.hexdigest()
                break
            nonce += 1

# Use shared memory for inter-thread communication.
# Ignore race conditions handling, 
# because it is essentially a 'race' between the thread
found = False
def mining(id, nthread, ldata):
    global found
    nonce = id
    while not found:
        s = ldata + str(nonce)
        result = hashlib.sha256(s.encode())
        if result.hexdigest()[:5] == prefix:
            global_result = result.hexdigest()
            break
        nonce += nthread

def parallel_multi_thread(nthread):
    threads = []
    for i in range(ntrial):
        found = False
        ldata = data + str(i)
        for i in range(nthread):
            t = threading.Thread(target=mining, args=(i, nthread, ldata))
            threads.append(t)
            t.start()
        
        for t in threads:
            t.join()

def parallel_multi_process(nproc):
    procs = []
    for i in range(ntrial):
        found = False
        ldata = data + str(i)
        for i in range(nproc):
            p = multiprocessing.Process(target=mining, args=(i, nproc, ldata))
            procs.append(p)
            p.start()
        
        for t in procs:
            p.join()
    
def main():
    print(f"difficult: {difficulty}, number of trials: {ntrial}")

    # sequential
    print("-" * 40)
    start = time.perf_counter()
    sequential()
    end = time.perf_counter()
    print(f"sequential version:")
    print(f"total time consumed: {end - start}")
    print(f"average time consumed: {(end - start) / ntrial}")
    print(f"last hash computed: {global_result}")

    # # multi-thread parallel
    print("-" * 40)
    nthread = 4
    start = time.perf_counter()
    parallel_multi_thread(nthread)
    end = time.perf_counter()
    print(f"{nthread} thread parallel version:")
    print(f"time consumed: {end - start}")
    print(f"average time consumed: {(end - start) / ntrial}")
    print(f"last hash computed: {global_result}")

    # multi-process parallel
    print("-" * 40)
    nproc = 8
    start = time.perf_counter()
    parallel_multi_process(nproc)
    end = time.perf_counter()
    print(f"{nproc} processes parallel version:")
    print(f"time consumed: {end - start}")
    print(f"average time consumed: {(end - start) / ntrial}")
    print(f"last hash computed: {global_result}")

if __name__ == "__main__":
    main()
